onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -122,11 +123,218 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
+  return Status::OK();
+}
+
+Status SequentialExecutor::ExecutePartial(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
+                                          const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
+                                          std::vector<OrtValue>& fetches, const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                          const logging::Logger& logger, int64_t run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  std::shared_ptr<ExecutionFrame> frame = nullptr;
+  std::map<int64_t, std::pair<std::unique_ptr<onnxruntime::ExecutionFrame>, size_t>>::iterator it;
+  size_t program_counter = 0;
+  //REVIEW(codemzs): We can get rid of this const cast by basically making partial graph exec manager APIs as SessionState.
+  PartialGraphExecutionManager& partial_graph_runs_manager = const_cast<SessionState&>(session_state).GetPartialGraphExecutionManager();
+  if (partial_graph_runs_manager.IsToBeExecutedRunId(run_id)) {
+    frame = std::shared_ptr<ExecutionFrame>(new ExecutionFrame(feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state));
+    partial_graph_runs_manager.PreparePartialRun(frame, run_id, program_counter);
+  } else {
+    auto run = partial_graph_runs_manager.GetPartialRun(run_id);
+    (run.first)->UpdateFeeds(feed_mlvalue_idxs, feeds);
+    (run.first)->UpdateFetches(fetch_mlvalue_idxs, fetches);
+    program_counter = run.second;
+    frame = run.first;
+  }
+
+  // Determine partial graph execution start and end nodes.
+  size_t program_counter_end;
+  for (program_counter_end = program_counter;
+       (program_counter_end < exec_plan_size) &&
+       ((session_state.GetKernel(exec_plan_vec[program_counter_end].node_index)->KernelDef().OpName() != "BreakOp") ||
+        (program_counter_end == program_counter));
+       program_counter_end += 1)
+    ;
+
+  ORT_RETURN_IF_ERROR(Execute(session_state, feeds, fetch_mlvalue_idxs, logger, *frame, program_counter, program_counter_end));
+
+  // Make sure intermediate outputs (Break Op node inputs in this case) are ready in the event they are being asynchronously computed.
+  if (program_counter_end < exec_plan_size) {
+    ORT_RETURN_IF_ERROR(SynchronizeNodeInputs(seq_exec_plan, program_counter_end, session_state, *frame, terminate_flag_, logger));
+  }
+
+  // Fetch the outputs.
+  ORT_RETURN_IF_ERROR(FetchOutputsFromExecutionFrame(*frame, session_state.GetTransferIntermidiateTensorOwnership(), fetches, logger));
+
+  // Update the program counter.
+  if (session_state.GetTransferIntermidiateTensorOwnership()) {
+    program_counter = program_counter_end + 1;
```

Comment:
Do we have any test coverage for this path? (As far as I can tell, `GetTransferIntermediateTensorOwnership()` always returns false in the current implementation. It's not clear why skipping a node in the plan is the right thing to do, and it's also unclear if we'd end up with `program_counter > exec_plan_size` - leading to an exception on L326.)

Reply:
You don't want to execute the node if the ownership of its input tensors has been transfered to PyTorch because after each node execution we free up the tensors that are no longer used downstream "starting" from that node, it is something we create as part of the memory plan. Of course there are otherways to tackle this problem but that was my rationale for skipping. I can add a test for this. 

---
In reply to: [600943925](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r600943925) [](ancestors = 600943925)